src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -257,11 +233,68 @@ private void DrainQueue()
             }
         }
 
-        private static Solution MergeChanges(Solution solution, Solution? mutatedSolution)
+        private async Task<RequestContext> CreateRequestContextAsync(QueueItem queueItem, CancellationToken cancellationToken)
         {
-            // TODO: Merge in changes to the solution that have been received from didChange LSP methods
-            // https://github.com/dotnet/roslyn/issues/45427
-            return mutatedSolution ?? solution;
+            // This method asks the solution provider for a solution, and then updates it based on the state of the world
+            // as we know it. You may look at this and think "why doesn't the solution provider keep track of the state
+            // then it can just provide the right solutions" and at first glance that is appealing, but there are two big
+            // benefits to this queue tracking documents separately:
+            //
+            // 1. Since this queue manages the execution of handlers, we can ensure that the document state tracker
+            //    is only used from mutating request handlers, or outside of any handlers, and therefore we know
+            //    that calls to it cannot overlap, and it doesn't have to worry about threading.
+            //
+            // 2. We specifically only want the updated solution at the start of a request. If the solution provider
+            //    handed out "the right" solution, then a request could ask it for something at the start of processing
+            //    and at the end of processing, and get different results each time.
+
+            // There are multiple possible solutions that we could be interested in, so we need to find the document
+            // first and then get the solution from there. If we're not given a document, this will return the default
+            // solution
+            var (document, solution) = _solutionProvider.GetDocumentAndSolution(queueItem.TextDocument, queueItem.ClientName);
+
+            // Now we can update the solution to represent the LSP view of the world, with any text changes we received
+            solution = await GetSolutionWithReplacedDocumentsAsync(solution, cancellationToken).ConfigureAwait(false);
+
+            // Now we need to get the document again, because if the request is for an open document, the original
+            // one we got might not be up to date
+            if (document is not null)
```

Comment:
should we change the solution provider to just return the documentId instead of the document?  that way we can avoid having an invalid document hanging around

Reply:
I've done the bare minimum for this here, which makes the code a little more clear, but also added a note to https://github.com/dotnet/roslyn/issues/48617 to follow up.